pkg/health/health.go

```diff
@@ -0,0 +1,160 @@
+package health
+
+import (
+	"context"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/metrics"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/util"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// DiagnosisResult is a possible result after a diagnosis.
+type DiagnosisResult int32
+
+// Possible diagnosis results.
+const (
+	Unknown      DiagnosisResult = 0
+	Inconclusive DiagnosisResult = 1
+	Healthy      DiagnosisResult = 2
+	Unhealthy    DiagnosisResult = 3
+)
+
+// Diagnosis is the information about the health of the revision.
+type Diagnosis struct {
+	OverallResult DiagnosisResult
+	CheckResults  []*CheckResult
+}
+
+// CheckResult is information about a metrics criteria check.
+type CheckResult struct {
+	Threshold     float64
+	ActualValue   float64
+	IsCriteriaMet bool
+}
+
+// Diagnose attempts to determine the health of a revision.
+//
+// If the minimum number of requests is not met, then health cannot be
+// determined and Diagnosis.EnoughRequests is set to false.
+//
+// Otherwise, all metrics criteria are checked to determine if the revision is
+// healthy.
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
```

Comment:
I’ll try to explain the context here:

1. `provider` is a metrics backend that lets you query certain type of metrics. However, we specify "filters" such as "service_name=X,revision=Y,project=Z", and those (under the covers) might be implemented differently. 

    though I agree, actually we could make that as part of initialization process to the MetricsProvider (and initialize a different `provider` for each svc/revision/project/...)

2. `offset` is for being able to specify "metrics for last X minutes" (for example Stackdriver actually doesn't support last X minutes, rather, we need to compute startDate/endDate inside its metric provider impl).

3. `minRequests` is something we will (over time) transition to an actual `config.Metric` criteria (just like latency) so ideally this will go away from this method signature.